File: Top_Google_Questions/299._Bulls_and_Cows/solution.py
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        # get bulls
        d1 = {}
        d2 = {}
        bulls = 0
        for i in range(len(secret)):
            if secret[i] == guess[i]:
                bulls += 1
            else:
                d1[secret[i]] = d1.get(secret[i], 0) + 1
                d2[guess[i]] = d2.get(guess[i], 0) + 1
        
        # get cows
        cows = 0
        for k, v in d1.items():
            if k in d2:
                cows += min(v, d2[k])
        
        return str(bulls) + 'A' + str(cows) + 'B'

File: Top_Google_Questions/299._Bulls_and_Cows/solution.js
/**
 * @param {string} secret
 * @param {string} guess
 * @return {string}
 */
var getHint = function(secret, guess) {
    // count bulls
    let bull = 0;
    let c1 = {};
    let c2 = {};
    for (let i=0; i<secret.length; i++) {
        c1[secret[i]] = c1.hasOwnProperty(secret[i]) ? c1[secret[i]] + 1 : 1;
        c2[guess[i]] = c2.hasOwnProperty(guess[i]) ? c2[guess[i]] + 1 : 1;
        if (secret[i] === guess[i]) {
            bull += 1;
        }
    }
    
    // count cows, then cows -= bulls
    let cow = 0;
    for (k of Object.keys(c1)) {
        if (c2.hasOwnProperty(k)) {
            cow += Math.min(c1[k], c2[k]);
        }
    }
    cow -= bull;
    
    return bull.toString() + 'A' + cow.toString() + 'B';
};